File: old/binary_search/search_range/search_range.py
import random


def binary_search_right_most_index(arr, val):
    left = 0
    right = len(arr) - 1
    right_most_index = -1

    while left <= right:
        middle = left + (right - left) // 2
        if arr[middle] < val:
            left = middle + 1
        elif arr[middle] > val:
            right = middle - 1
        else:
            right_most_index = middle
            left = middle + 1

    return right_most_index


def binary_search_left_most_index(arr, val):
    left = 0
    right = len(arr) - 1
    left_most_index = -1

    while left <= right:
        middle = left + (right - left) // 2
        if arr[middle] < val:
            left = middle + 1
        elif arr[middle] > val:
            right = middle - 1
        else:
            left_most_index = middle
            right = middle - 1

    return left_most_index


def main():
    input = [0, 1, 1, 1, 2, 3, 3, 3, 3, 4, 5, 5, 5, 5, 5, 5, 6, 7, 8, 8, 10]
    print(input)
    find_val = random.randint(0, 10)
    print(f"we are finding: {find_val}")
    left_index = binary_search_left_most_index(input, find_val)
    right_index = binary_search_right_most_index(input, find_val)
    if left_index != -1 and right_index != -1:
        print(f"[{left_index}, {right_index}]")
        return
    elif left_index != -1:
        print(f"[{left_index}, {left_index}]")
        return
    else:
        print(f"[{right_index}, {right_index}]")
        return


main()

File: old/binary_search/search_range/search_range.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>


using namespace std;

int binarySearchLeftMostIndex(vector<int>& arr, int val) {
    int left = 0;
    int right = arr.size() - 1;
    int leftMostIndex = -1;

    while (left <= right) {
        int middle = left + (right - left) / 2;
        if (arr[middle] < val) {
            left = middle + 1;
        } else if (arr[middle] > val) {
            right = middle - 1;
        } else {
            leftMostIndex = middle;
            right = middle - 1;
        } 
    }


    return leftMostIndex;
}

int binarySearchRightMostIndex(vector<int>& arr, int val) {
    int left = 0;
    int right = arr.size() - 1;
    int rightMostIndex = -1;

    while (left <= right) {
        int middle = left + (right - left) / 2;
        if (arr[middle] < val) {
            left = middle + 1;
        } else if (arr[middle] > val) {
            right = middle - 1;
        } else {
            rightMostIndex = middle;
            left = middle + 1;
        } 
    }


    return rightMostIndex;
}

int main() {
    srand(time(0));

    vector<int> input = {0, 0, 0, 1, 1, 2, 3, 4, 4, 4, 5, 5, 5, 5, 5, 6, 7, 8, 10};
    for (int val: input) {
        cout << val << " ";
    }
    cout << endl;

    int findVal = rand() % 11; // from 0 to 11
    
    cout << "we are looking for: " << findVal << endl;


    int leftIndex = binarySearchLeftMostIndex(input, findVal);
    int rightIndex = binarySearchRightMostIndex(input, findVal);

    if (leftIndex == -1) {
        cout << "[-1, ";
    } else {
        cout << "[" << leftIndex << ", ";
    } 

         
    if (rightIndex == -1) {
        cout << "-1]"; 
    } else {
        cout << rightIndex << "]";
    }

    return 0;
}
